Sources/NIOConcurrencyHelpers/fastatomics.swift

```diff
@@ -0,0 +1,269 @@
+    public static let fast_atomic_compare_and_exchange         = catmc_fast_atomic_unsigned_long_compare_and_exchange
+    public static let fast_atomic_add                          = catmc_fast_atomic_unsigned_long_add
+    public static let fast_atomic_sub                          = catmc_fast_atomic_unsigned_long_sub
+    public static let fast_atomic_exchange                     = catmc_fast_atomic_unsigned_long_exchange
+    public static let fast_atomic_load                         = catmc_fast_atomic_unsigned_long_load
+    public static let fast_atomic_store                        = catmc_fast_atomic_unsigned_long_store
+}
+
+public class FastAtomic<T: FastAtomicPrimitive> {
+    @usableFromInline
+    typealias Manager = ManagedBufferPointer<Void, T.AtomicWrapper>
+
+    /// Create an atomic object with `value`
+    @inlinable
+    static func makeAtomic(value: T) -> FastAtomic {
+        let manager = Manager(bufferClass: self, minimumCapacity: 1) { _, _ in }
+        manager.withUnsafeMutablePointerToElements {
+            T.fast_atomic_create_with_existing_storage($0, value)
+        }
+        return manager.buffer as! FastAtomic<T>
+    }
+
+    /// Atomically compares the value against `expected` and, if they are equal,
+    /// replaces the value with `desired`.
+    ///
+    /// This implementation conforms to C11's `atomic_compare_exchange_strong`. This
+    /// means that the compare-and-swap will always succeed if `expected` is equal to
+    /// value. Additionally, it uses a *sequentially consistent ordering*. For more
+    /// details on atomic memory models, check the documentation for C11's
+    /// `stdatomic.h`.
+    ///
+    /// - Parameter expected: The value that this object must currently hold for the
+    ///     compare-and-swap to succeed.
+    /// - Parameter desired: The new value that this object will hold if the compare
+    ///     succeeds.
+    /// - Returns: `True` if the exchange occurred, or `False` if `expected` did not
+    ///     match the current value and so no exchange occurred.
+    @inlinable
+    public func compareAndExchange(expected: T, desired: T) -> Bool {
+        return Manager(unsafeBufferObject: self).withUnsafeMutablePointerToElements {
+            return T.fast_atomic_compare_and_exchange($0, expected, desired)
+        }
+    }
+
+    /// Atomically adds `rhs` to this object.
+    ///
+    /// This implementation uses a *relaxed* memory ordering. This guarantees nothing
+    /// more than that this operation is atomic: there is no guarantee that any other
+    /// event will be ordered before or after this one.
+    ///
+    /// - Parameter rhs: The value to add to this object.
+    /// - Returns: The previous value of this object, before the addition occurred.
+    @inlinable
+    public func add(_ rhs: T) -> T {
+        return Manager(unsafeBufferObject: self).withUnsafeMutablePointerToElements {
+            return T.fast_atomic_add($0, rhs)
```

Comment:
@atrick just to double check, we can rely on the tail allocated storage in our class to always have a const address, right?

Reply:
Yeah, there's no way to move a class object. IIRC the way `realloc` would work is by providing a new reference (old references will never see the realloc'd class).